codice/ddf
# DDF-3908 Upgrade to Karaf 4.2.0

#### What does this PR do?

Upgrades from Karaf 4.1.2 to Karaf 4.2.0.
Major Notable Changes in Karaf:

- Slimmed down distribution by reducing the number of libraries built into the distribution under `lib/`. (https://github.com/apache/karaf/commit/e7ca1e209f02d844a481bb5d95a351aba47c8c6c#diff-564886f3f54fbea9414b6d9c0fefc315L193) DDF relies on some of those libraries, primarily for various XML operations, so these needed to be added back in.
- Karaf tightened up how they resolve features, which primarily impacted DDF's feature itests.
- Bumped a minor version in Jetty which has significant changes in their session architecture, which forces some changes in how DDF handles sessions.

#### Who is reviewing it? 
@emmberk @paouelle @tbatie @bakejeyner @blen-desta @ryeats 
#### Select relevant component teams: 
https://github.com/orgs/codice/teams/security
#### Ask 2 committers to review/merge the PR and tag them here.
@coyotesqrl
@lessarderic
@stustison
@clockard
@vinamartin
@bdeining 

#### How should this be tested?
Across Windows and *Nix OS's:
- Manually test basic functionality of DDF.
- Manually modify some properties in custom.system.properties, restart, and verify the properties are updated.
- Drop .cfg and .config files in the /etc directory, and confirm that the configurations are installed successfully.
- Confirm that the bin scripts work as expected.
- Walkthrough the Hardening Guide.
- Confirm that the waitForReady command still works.
- Confirm that installing via each install profile works.
- Confirm that logging still works and is still configurable from org.ops4j.pax.logging.cfg and from the Karaf console. Also, try switching the logging configuration to the xml-style configuration (see comment in org.ops4j.pax.logging.cfg, and confirm that logging still works and is configurable from log4j2.config.xml and from the Karaf console.
- Test that the common console commands are still functional.
- Confirm SSHing into a running instance still works. Confirm SSHing into itests still works.
- Confirm debugging a live system
- Confirm that the Karaf instance can be restarted.
- Confirm that the migration:export command.
- Test that profile console commands still work.
- Test that install with ?dev=true and restart still works 
- Test Automatic Start on System Boot (see Documentation)
- Test using the up and down arrows on the console command line
- Test  log:tail, log:set <level> <logger>, log:get, log:list
- Test that setting the karaf.restart.jvm property to true causes system halt commands to behave as reboots
    - set karaf.restart.jvm to true
    - run the reboot command
    - it should shutdown and restart
#### Any background context you want to provide?
#### What are the relevant tickets?
[DDF-3908](https://codice.atlassian.net/browse/DDF-3908)
#### Screenshots

#### Checklist:

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
